core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/entity/decorator/FilteringEntityByEntitySelector.java

```diff
@@ -0,0 +1,411 @@
+                    return entity;
+                }
+            }
+            return null;
+        }
+
+        private Object pickSelected() {
+            if (selected == null) {
+                throw new NoSuchElementException();
+            }
+            var result = selected;
+            this.selected = null;
+            return result;
+        }
+
+        @Override
+        public Object next() {
+            return pickSelected();
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            this.selected = pickPrevious();
+            return selected != null;
+        }
+
+        private Object pickPrevious() {
+            if (selected != null) {
+                throw new IllegalStateException("The next value has already been picked.");
+            }
+            initialize();
+            this.selected = null;
+            while (entityIterator.hasPrevious()) {
+                var entity = entityIterator.previous();
+                if (isReachable(entity)) {
+                    return entity;
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public Object previous() {
+            return pickSelected();
+        }
+
+        @Override
+        public int nextIndex() {
+            return entityIterator.nextIndex();
+        }
+
+        @Override
+        public int previousIndex() {
+            return entityIterator.previousIndex();
+        }
+
+        @Override
+        public void remove() {
+            entityIterator.remove();
+        }
+
+        @Override
+        public void set(Object o) {
+            entityIterator.set(o);
+        }
+
+        @Override
+        public void add(Object o) {
+            entityIterator.add(o);
+        }
+    }
+
+    private static class RandomFilteringValueRangeIterator<Solution_> extends AbstractFilteringValueRangeIterator<Solution_> {
+
+        private final Iterator<Object> entityIterator;
+        private final int maxBailoutSize;
+
+        private RandomFilteringValueRangeIterator(Supplier<Object> upcomingEntitySupplier, Iterator<Object> entityIterator,
+                BasicVariableDescriptor<Solution_>[] basicVariableDescriptors, ValueRangeManager<Solution_> valueRangeManager,
+                int maxBailoutSize) {
+            super(upcomingEntitySupplier, basicVariableDescriptors, valueRangeManager);
+            this.entityIterator = entityIterator;
+            this.maxBailoutSize = maxBailoutSize;
+        }
+
+        @Override
+        public boolean hasNext() {
+            initialize();
+            return entityIterator.hasNext();
+        }
+
+        @Override
+        public Object next() {
+            initialize();
+            if (!entityIterator.hasNext()) {
+                // If no reachable entity is found, we return the currently selected entity,
+                // which will result in a non-doable move
+                return currentReplayedEntity();
```

Comment:
Not too excited here. Typically in this situation, we return `noUpcomingSelection()`, not an invalid value.